### PR TITLE
[Azure Function App] Uploading package and generating SAS URL

### PR DESCRIPTION
**Pull request by 20shivangi**

### CODE REVIEW

File: Tasks/AzureFunctionAppV1/package-lock.json [hunk 1]
```diff
@@ -24,6 +24,16 @@
       "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.0.9.tgz",
       "integrity": "sha1-kbR5JYinc4wl813W9jdSovh3YTU="
     },
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/package-lock.json]: I will have the AI on me to polish this file.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 2]
```diff
@@ -30,6 +38,71 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+                    }
+                };
+            
+                let token = blobService.generateSharedAccessSignature(containerName, blobName, sharedAccessPolicy);
+                let sasUrl = blobService.getUrl(containerName, blobName, token);
+                tl.debug(`SAS URL is: ${sasUrl}`);
+                defer.resolve(sasUrl);
+            });
+        });
+        return defer.promise;
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: new line

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 3]
```diff
@@ -30,6 +38,71 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+    }
+
+    private async publishRunFromPackage(sasUrl) {
+        await this.appService.patchApplicationSettings({'WEBSITE_RUN_FROM_PACKAGE': sasUrl});
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: Can we have a console statement here to say that we are updating the app settings?

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 4]
```diff
@@ -30,6 +38,71 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
         }
         return storageData;
     }
+
+    private async uploadPackage(storageDetails, deployPackage) : Promise<string> {
+        var defer = Q.defer<string>();
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: `var` to `let`

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 5]
```diff
@@ -30,6 +38,71 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+        let storageAccount = storageDetails["AccountName"];
+        let storageKey = storageDetails["AccountKey"];
+        const blobService = azureStorage.createBlobService(storageAccount, storageKey);
+
+        const containerName = 'azure-pipelines-deploy';
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: specify `type` wherever possible.
` const containerName: string = 'azure-pipelines-deploy';`

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 6]
```diff
@@ -30,6 +38,71 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+        const blobName = `package_${Date.now()}.zip`;
+        let fileName;
+
+        switch(deployPackage.getPackageType()){
+            case PackageType.folder:
+                let tempPackagePath = webCommonUtility.generateTemporaryFolderOrZipPath(tl.getVariable('AGENT.TEMPDIRECTORY'), false);
+                let archivedWebPackage = await zipUtility.archiveFolder(deployPackage.getPath(), "", tempPackagePath);
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: have a try catch here.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 2 above)
Comment (20shivangi) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: It is done.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 4 above)
Comment (20shivangi) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: It is done.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 5 above)
Comment (20shivangi) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: It is done

**Comment by vincentdass** [comment]
change the statement `old value : ${applicationSettings.properties[key]}. new value: ${addProperties[key]}` to `value for ${key} has been changed`

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 7]
```diff
@@ -30,6 +38,71 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+                    AccessPolicy: {
+                        Permissions: azureStorage.BlobUtilities.SharedAccessPermissions.READ,
+                        Start: startDate,
+                        Expiry: expiryDate
+                    }
+                };
+            
+                let token = blobService.generateSharedAccessSignature(containerName, blobName, sharedAccessPolicy);
+                let sasUrl = blobService.getUrl(containerName, blobName, token);
+                tl.debug(`SAS URL is: ${sasUrl}`);
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: make `sasUrl` query parameters as secret variable.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 3 above)
Comment (vincentdass) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: console.log(tl.loc('UpdatedRunFromPackageSettings', sasUrl)
UpdatingRunFromPackageSettings: "Updated 'W_R_F_P Application setting to %s"

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 8]
```diff
@@ -30,6 +38,71 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+                defer.resolve(sasUrl);
+            });
+        });
+        return defer.promise;
+    }
+
+    private async publishRunFromPackage(sasUrl) {
+        await this.appService.patchApplicationSettings({'WEBSITE_RUN_FROM_PACKAGE': sasUrl});
+        await sleepFor(5);
```
Comment (SumiranAgg) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: Why do we sleep here. Shouldnt we just rely on a 200 response from patch app settings call and call sync trigger?

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 3 above)
Comment (20shivangi) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: It is done.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 6 above)
Comment (20shivangi) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: It is done.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 8 above)
Comment (vincentdass) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: we follow the same stpes of [functions-core-tools](https://github.com/Azure/azure-functions-core-tools/blob/dab81f32ae4405c8f747efa23eb18273b6f98a51/src/Azure.Functions.Cli/Actions/AzureActions/PublishFunctionAppAction.cs#L312).

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 7 above)
Comment (20shivangi) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: It is done.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 9]
```diff
@@ -30,6 +38,83 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+                throw new Error(tl.loc('Invalidwebapppackageorfolderpathprovided', deployPackage.getPath()));
+        }
+
+        blobService.createContainerIfNotExists(containerName, error => {
+            if (error){
+                defer.reject(error);
+            }
+
+            //upoading package
+            blobService.createBlockBlobFromLocalFile(containerName, blobName, fileName, (error, result) => {
```
Comment (vincentdass) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: `azure-storage` has its own retry logic: https://github.com/Azure/azure-storage-node/blob/66d24e89eeedf77ab83356c69fe5a39ed03b73b9/lib/common/filters/retrypolicyfilter.js

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 10]
```diff
@@ -30,6 +38,83 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
     }
+
+    private async uploadPackage(storageDetails, deployPackage) : Promise<string> {
+        let defer = Q.defer<string>();
+        let storageAccount = storageDetails["AccountName"];
+        let storageKey = storageDetails["AccountKey"];
+        const blobService = azureStorage.createBlobService(storageAccount, storageKey);
+
+        const containerName: string = 'azure-pipelines-deploy';
+        const blobName: string = `package_${Date.now()}.zip`;
```
Comment (kmkumaran) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: Can this vary based on agent locale and will it continue to have supported character set for blob name?

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts [hunk 11]
```diff
@@ -30,6 +38,83 @@ export class ConsumptionWebAppDeploymentProvider extends AzureRmWebAppDeployment
...
+            //upoading package
+            blobService.createBlockBlobFromLocalFile(containerName, blobName, fileName, (error, result) => {
+                if (error) {
+                    defer.reject(error);
+                }
+
+                //generating SAS URL
+                let startDate = new Date();
+                let expiryDate = new Date(startDate);
+                expiryDate.setFullYear(startDate.getUTCFullYear() + 1);
```
Comment (kmkumaran) [inline Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: Can we find what Az cli is doing in this case?

Comment (kmkumaran) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: should user have choice to configure the expiry date?

File: Tasks/AzureFunctionAppV1/package.json [hunk 12]
```diff
@@ -20,6 +20,7 @@
     "@types/mocha": "2.2.48",
     "@types/node": "6.0.68",
     "@types/q": "1.0.7",
+    "azure-storage": "2.2.1",
```
[the remark this answers is not in the document]
Comment (vincentdass) [reply Tasks/AzureFunctionAppV1/package.json]: `azure-storage` component is already OSS approved and IP scanned. so ideally all the nested components should have been IP scanned. I will hve AI on me to polish the package-lock file

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 10 above)
Comment (vincentdass) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: `Date.now()` would return a timestamp of type `int`. we should be good here.

File: Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts (hunk 11 above)
Comment (vincentdass) [reply Tasks/AzureFunctionAppV1/deploymentProvider/ConsumptionWebAppDeploymentProvider.ts]: Azure Functions core tools has an expitry time of [10 years](https://github.com/Azure/azure-functions-core-tools/blob/dab81f32ae4405c8f747efa23eb18273b6f98a51/src/Azure.Functions.Cli/Actions/AzureActions/PublishFunctionAppAction.cs#L450). Will check with Web App team on this.

Regarding the second question, I had the same thought of exposing it as secondary variable (release variable). But, planned to accomodate once customers ask for it.

**Pull request merged**